File: C++/SPHAmod.cpp
#include "sphamod.h"
#include <bitset>

using namespace std;

spha::spha(int x, int y, int z) : x(x), y(y), z(z) {}

spha::spha(const vector<long double>& initialData) : data(initialData) {}

vector<long double> spha::concatenate(const vector<long double>& a, const vector<long double>& b) const {
    vector<long double> result = a;
    result.insert(result.end(), b.begin(), b.end());
    return result;
}

vector<long double> spha::slice(const vector<long double>& v, int start, int end) const {
    return vector<long double>(v.begin() + start, v.begin() + end);
}

int spha::ps(int index) const {
    if (index >= (1 << (x+y))) return 0;
    string binary = bitset<32>(index).to_string().substr(32 - (x+y+z));
    int ones_count = count(binary.begin() + y + z, binary.end(), '1');
    return (ones_count % 2 == 0) ? 1 : -1;
}

vector<long double> spha::md(const vector<long double>& args) const {
    return vector<long double>(args.begin(), args.begin() + pow(2, x + y + z));
}

vector<long double> spha::md1(const vector<long double>& a, const vector<long double>& b) const {
    int n = x + y + z;
    if (n == 0) return {a[0] * b[0]};
    if (n == 1) {
        if (x == 1) return {a[0]*b[0] - a[1]*b[1], a[0]*b[1] + a[1]*b[0]};
        else if (y == 1) return {a[0]*b[0] + a[1]*b[1], a[0]*b[1] + a[1]*b[0]};
        else return {a[0]*b[0], a[0]*b[1] + a[1]*b[0]};
    } else {
        int half = pow(2, n-1);
        vector<long double> a1 = slice(a, 0, half), a2 = slice(a, half, a.size());
        vector<long double> b1 = slice(b, 0, half), b2 = slice(b, half, b.size());
        vector<long double> c1, c2;
        if (z > 0) {
            spha subspha(x, y, z-1);
            c1 = subspha.md1(a1, b1);
            c2 = subspha.md1(a1, b2);
            transform(c2.begin(), c2.end(), subspha.md1(a2, b1).begin(), c2.begin(), plus<long double>());
        } else if (y > 0) {
            spha subspha(x, y-1, z);
            c1 = subspha.md1(a1, b1);
            transform(c1.begin(), c1.end(), subspha.md1(a2, b2).begin(), c1.begin(), plus<long double>());
            c2 = subspha.md1(a1, b2);
            transform(c2.begin(), c2.end(), subspha.md1(a2, b1).begin(), c2.begin(), plus<long double>());
        } else {
            spha subspha(x-1, y, z);
            c1 = subspha.md1(a1, b1);
            transform(c1.begin(), c1.end(), subspha.md1(a2, b2).begin(), c1.begin(), minus<long double>());
            c2 = subspha.md1(a1, b2);
            transform(c2.begin(), c2.end(), subspha.md1(a2, b1).begin(), c2.begin(), plus<long double>());
        }
        return concatenate(c1, c2);
    }
}

vector<long double> spha::md2(const vector<long double>& a, const vector<long double>& b) const {
    int n = x + y + z;
    if (n == 0) return {a[0] / b[0]};
    if (n == 1) {
        if (x == 1) {
            long double denominator = b[0]*b[0] + b[1]*b[1];
            return {(a[0]*b[0] + a[1]*b[1]) / denominator, (a[1]*b[0] - a[0]*b[1]) / denominator};
        } else if (y == 1) {
            long double denominator = b[0]*b[0] - b[1]*b[1];
            return {(a[0]*b[0] - a[1]*b[1]) / denominator, (a[1]*b[0] - a[0]*b[1]) / denominator};
        } else {
            return {a[0]*b[0] / (b[0]*b[0]), (a[1]*b[0] - a[0]*b[1]) / (b[0]*b[0])};
        }
    } else {
        int half = pow(2, n-1);
        vector<long double> b1 = slice(b, 0, half);
        vector<long double> b2 = slice(b, half, b.size());
        vector<long double> negB2(b2.size());
        transform(b2.begin(), b2.end(), negB2.begin(), negate<long double>());
        vector<long double> numerator = md1(a, concatenate(b1, negB2));
        vector<long double> denominator;
        if (z > 0) {
            spha subspha(x, y, z-1);
            denominator = subspha.md1(b1, b1);
            return concatenate(
                subspha.md2(slice(numerator, 0, half), denominator),
                subspha.md2(slice(numerator, half, numerator.size()), denominator)
            );
        } else if (y > 0) {
            spha subspha(x, y-1, z);
            denominator = subspha.md1(b1, b1);
            transform(denominator.begin(), denominator.end(), subspha.md1(b2, b2).begin(), denominator.begin(), minus<long double>());
            return concatenate(
                subspha.md2(slice(numerator, 0, half), denominator),
                subspha.md2(slice(numerator, half, numerator.size()), denominator)
            );
        } else {
            spha subspha(x-1, y, z);
            denominator = subspha.md1(b1, b1);
            transform(denominator.begin(), denominator.end(), subspha.md1(b2, b2).begin(), denominator.begin(), plus<long double>());
            return concatenate(
                subspha.md2(slice(numerator, 0, half), denominator),
                subspha.md2(slice(numerator, half, numerator.size()), denominator)
            );
        }
    }
}

vector<long double> spha::mp(long double a, const vector<long double>& b) const {
    int n = x + y + z;
    vector<long double> result(pow(2, n), 0);

    if (a != 0) {
        long double abs_a = abs(a);
        long double log_abs_a = log(abs_a);

        if (a > 0) {
            result[0] = pow(a, b[0]);
        } else {
            result[0] = pow(-a, b[0]) * cos(b[0] * M_PI);
            result[pow(2, n-1)] = pow(-a, b[0]) * sin(b[0] * M_PI);
        }

        for (int i = 1; i < pow(2, n); i++) {
            vector<long double> term(pow(2, n), 0);
            int ps_i = ps(i);

            if (ps_i > 0) {
                term[0] = cosh(b[i] * log_abs_a);
                term[i] = sinh(b[i] * log_abs_a);
            } else if (ps_i < 0) {
                term[0] = cos(b[i] * log_abs_a);
                term[i] = sin(b[i] * log_abs_a);
            } else {
                term[0] = 1;
                term[i] = b[i] * log_abs_a;
            }

            result = md1(result, term);

            if (a < 0) {
                long double rotationFactor = exp(ps_i * M_PI * b[i]);
                for (auto& v : result) {
                    v *= rotationFactor;
                }
            }
        }
    } else {
        if (all_of(b.begin(), b.end(), [](long double v) { return v == 0; })) {
            result[0] = 1;
        } else {
            result[0] = 0;
        }
    }

    return result;
}

spha& spha::operator=(const spha& other) {
    if (this != &other) {
        if (data.size() != other.data.size()) {
            throw invalid_argument("Vectors must be of the same size");
        }
        for (size_t i = 0; i < data.size(); ++i) {
            data[i] = other.data[i];
        }
    }
    return *this;
}

vector<long double> operator+(const vector<long double>& a, const vector<long double>& b) {
    if (a.size() != b.size()) {
        throw invalid_argument("Vectors must be of the same size");
    }
    vector<long double> result(a.size());
    for (size_t i = 0; i < a.size(); ++i) {
        result[i] = a[i] + b[i];
    }
    return result;
}

vector<long double> operator-(const vector<long double>& a, const vector<long double>& b) {
    if (a.size() != b.size()) {
        throw invalid_argument("Vectors must be of the same size");
    }
    vector<long double> result(a.size());
    for (size_t i = 0; i < a.size(); ++i) {
        result[i] = a[i] - b[i];
    }
    return result;
}

vector<long double> operator*(const vector<long double>& a, long double scalar) {
    vector<long double> result(a.size());
    for (size_t i = 0; i < a.size(); ++i) {
        result[i] = a[i] * scalar;
    }
    return result;
}

vector<long double> operator*(long double scalar, const vector<long double>& a) {
    return a * scalar;
}

vector<long double>& operator+=(vector<long double>& a, const vector<long double>& b) {
    if (a.size() != b.size()) {
        throw invalid_argument("Vectors must be of the same size");
    }
    for (size_t i = 0; i < a.size(); ++i) {
        a[i] += b[i];
    }
    return a;
}

vector<long double>& operator-=(vector<long double>& a, const vector<long double>& b) {
    if (a.size() != b.size()) {
        throw invalid_argument("Vectors must be of the same size");
    }
    for (size_t i = 0; i < a.size(); ++i) {
        a[i] -= b[i];
    }
    return a;
}

vector<long double>& operator*=(vector<long double>& a, long double scalar) {
    for (size_t i = 0; i < a.size(); ++i) {
        a[i] *= scalar;
    }
    return a;
}

ostream& operator<<(ostream& out, const vector<long double>& a) {
    out << setprecision(15);
    out << a[0];
    for (size_t i = 1; i < a.size(); ++i) {
        out << ", " << a[i];
    }
    return out;
}

File: Python/SPHAmod.py
#The iteration order is E=>D=>C
import numpy as np
from math import cosh,sinh,cos,sin,log,e,pi
def md(x, y, z):
    """Creating SPHA"""
    return lambda *args: np.array(args[:2**(x+y+z)])
def md1(x, y, z):
    """SPHA Multiplication"""
    def mul(a, b):
        n = x + y + z
        if n == 0 : return np.array([a[0]*b[0]])
        if n == 1:
            if x == 1 : return np.array([a[0]*b[0] - a[1]*b[1], a[0]*b[1] + a[1]*b[0]])
            elif y == 1 : return np.array([a[0]*b[0] + a[1]*b[1], a[0]*b[1] + a[1]*b[0]])
            else : return np.array([a[0]*b[0], a[0]*b[1] + a[1]*b[0]])
        else:
            a1, a2, b1, b2= a[:2**(n-1)], a[2**(n-1):], b[:2**(n-1)], b[2**(n-1):]
            if z > 0:
                c1 = md1(x, y, z-1)(a1, b1)
                c2 = md1(x, y, z-1)(a1, b2) + md1(x, y, z-1)(a2, b1)
            elif y > 0:
                c1 = md1(x, y-1, z)(a1, b1) + md1(x, y-1, z)(a2, b2)
                c2 = md1(x, y-1, z)(a1, b2) + md1(x, y-1, z)(a2, b1)
            else:
                c1 = md1(x-1, y, z)(a1, b1) - md1(x-1, y, z)(a2, b2)
                c2 = md1(x-1, y, z)(a1, b2) + md1(x-1, y, z)(a2, b1)
            return np.concatenate((c1, c2))
    return mul
def md2(x, y, z):
    """SPHA Division"""
    def div(a, b):
        n = x + y + z
        if n == 0 : return np.array([a[0]/b[0]])
        if n == 1:
            if x == 1:
                denominator = b[0]**2 + b[1]**2
                return np.array([(a[0]*b[0] + a[1]*b[1]) / denominator,(a[1]*b[0] - a[0]*b[1]) / denominator])
            elif y == 1:
                denominator = b[0]**2 - b[1]**2
                return np.array([(a[0]*b[0] - a[1]*b[1]) / denominator,(a[1]*b[0] - a[0]*b[1]) / denominator])
            else : return np.array([a[0]*b[0] / b[0]**2,(a[1]*b[0] - a[0]*b[1]) / b[0]**2])
        else:
            b1 = b[:2**(n-1)]
            b2 = b[2**(n-1):]
            numerator = md1(x, y, z)(a, np.concatenate((b1, -b2)))
            if z > 0:
                denominator = md1(x, y, z-1)(b1, b1)
                return np.concatenate((md2(x, y, z-1)(numerator[:2**(n-1)], denominator), md2(x, y, z-1)(numerator[2**(n-1):], denominator)))
            elif y > 0:
                denominator = md1(x, y-1, z)(b1, b1) - md1(x, y-1, z)(b2, b2)
                return np.concatenate((md2(x, y-1, z)(numerator[:2**(n-1)], denominator), md2(x, y-1, z)(numerator[2**(n-1):], denominator)))
            else:
                denominator = md1(x-1, y, z)(b1, b1) + md1(x-1, y, z)(b2, b2)
                return np.concatenate((md2(x-1, y, z)(numerator[:2**(n-1)], denominator), md2(x-1, y, z)(numerator[2**(n-1):], denominator)))
    return div
def ps(index, x, y, z):#informal function
    if index >= 2**(x+y): return 0
    binary = bin(index)[2:].zfill(x+y+z)[y+z:]
    ones_count = binary.count('1')
    return (ones_count % 2 == 0)*2-1
def mp(x, y, z):
    """SPHA Exponentiation"""
    def pow(a, b):
        n = x + y + z
        result = np.zeros(2**n)
        if a != 0:
            abs_a = abs(a)
            log_abs_a = log(abs_a)
            if a > 0 : result[0] = a**b[0]
            else:
                result[0] = abs_a**b[0] * cos(b[0]*pi)
                result[2**(n-1)] = abs_a**b[0] * sin(b[0]*pi)
            for i in range(1, 2**n):
                term = np.zeros(2**n)
                ps_i = ps(i, x, y, z)
                if ps_i > 0:
                    term[0] = cosh(b[i] * log_abs_a)
                    term[i] = sinh(b[i] * log_abs_a)
                elif ps_i < 0:
                    term[0] = cos(b[i] * log_abs_a)
                    term[i] = sin(b[i] * log_abs_a)
                else:
                    term[0] = 1
                    term[i] = b[i] * log_abs_a
                result = md1(x, y, z)(result, term)
                if a < 0 : result *= e**(ps_i * pi * b[i])
        else:
            if np.all(b == 0) : result[0] = 1
        return result
    return pow
